impeller/docs/faq.md

```diff
@@ -100,3 +100,133 @@
+    the entire set of shaders an application might need. So, applications would
+    create and compile these shaders just-in-time. Once they were compiled, they
+    could be cached.
+  * But this led to the problem of jank during the initial run of performance
+    sensitive areas of the application when there were no cached shaders.
+  * Fortunately, early on, the team was able to predict the common shaders the
+    application was likely going to need. A generic ["warm
+    up"](https://github.com/flutter/flutter/pull/27660) phase was added where
+    these shaders were given a chance to be compiled and cached at the launch of
+    a Flutter application.
+  * As Flutter applications became more complex, the generic shader warmup was
+    no longer suitable for all apps. So the ability to specify shaders to warm
+    up was [exposed to end
+    users](https://api.flutter.dev/flutter/painting/ShaderWarmUp-class.html).
+  * But, this was exceedingly hard to do and also platform specific (iOS Metal,
+    Android OpenGL, Fuchsia Vulkan, etc.). So much so that only a handful of
+    engineers (who all worked on the Flutter Engine) could effectively build the
+    rendering intent that would warm up the right shaders. And these shaders
+    were specific to the app version. If the app was updated to add or remove
+    more features, the warmup routines were invalid and mostly just served to
+    slow down application startup.
+  * To make this process self-service, [a scheme was
+    devised](https://github.com/flutter/flutter/issues/53607) to run the
+    application and its tests during development in training phase, capture the
+    shaders, package them with the application, and during a new "warm up"
+    phase,  pre-compile the shaders.
+  * Remember, these shaders were platform specific. The scheme detailed above
+    unfortunately moved the burden to developers. It also made the application
+    sizes a lot bigger since the shaders now had to be packaged with the
+    application. Application launch times were delayed as well.
+  * Even developers who went through these steps weren't always guaranteed
+    freedom from shader compilation jank. If some screens (and thus, their
+    shaders) were missed during training runs due to incomplete automated or
+    manual testing, jank could be experienced. For complex applications where
+    certain flows were blocked behind authentication or other restrictions, this
+    was an easy situation to get into. Also, some animation frames could be
+    skipped during training because jank during previous frames in the training
+    run. So training runs not only had to be run over the entire app, but also
+    multiple times (on all platforms) to ensure success. Moreover, applications
+    that were really effective at training began to run into egregiously high
+    first-frame times ([some as high as 6
+    seconds](http://github.com/flutter/flutter/issues/97884)).
```

Comment:
Might be worth mentioning that this never worked at all on Android, due to the fact that skia generated different shaders based on GPU model

Reply:
Yeah, I remembered but got lazy and didn't want to look up the bugs to cross reference here. But https://github.com/flutter/flutter/issues/102655 seems to be it. Still open! Will add.

Reply:
Done. Also added a ref to things we tried to reduce shader variants.